src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -1390,52 +1390,24 @@ JvmtiEnvBase::get_object_monitor_usage(JavaThread* calling_thread, jobject objec
 
   uint32_t debug_bits = 0;
   // first derive the object's owner and entry_count (if any)
-  {
-    address owner = NULL;
-    {
-      markWord mark = hobj()->mark();
-
-      if (!mark.has_monitor()) {
-        // this object has a lightweight monitor
+  owning_thread = ObjectSynchronizer::get_lock_owner(tlh.list(), hobj);
+  if (owning_thread != NULL) {
+    Handle th(current_thread, get_vthread_or_thread_oop(owning_thread));
+    ret.owner = (jthread)jni_reference(calling_thread, th);
 
-        if (mark.has_locker()) {
-          owner = (address)mark.locker(); // save the address of the Lock word
-        }
-        // implied else: no owner
-      } else {
-        // this object has a heavyweight monitor
-        mon = mark.monitor();
-
-        // The owner field of a heavyweight monitor may be NULL for no
-        // owner, a JavaThread * or it may still be the address of the
-        // Lock word in a JavaThread's stack. A monitor can be inflated
-        // by a non-owning JavaThread, but only the owning JavaThread
-        // can change the owner field from the Lock word to the
-        // JavaThread * and it may not have done that yet.
-        owner = (address)mon->owner();
-      }
-    }
-
-    if (owner != NULL) {
-      // This monitor is owned so we have to find the owning JavaThread.
-      owning_thread = Threads::owning_thread_from_monitor_owner(tlh.list(), owner);
-      assert(owning_thread != NULL, "owning JavaThread must not be NULL");
```

Comment:
I finished doing an equivalence analysis of this code that has been replaced with
the code in `ObjectSynchronizer::get_lock_owner()`. This `assert()` on L1422 is
the only thing I found that is "missing".

However, I don't think that you can add back that `assert()` call in this function.
The reason that the `assert()` is okay in the original code is because it is
"protected" by this line:
```
L1419      if (owner != NULL) {
```
and while that check is also done in `ObjectSynchronizer::get_lock_owner()` on
L1032, it DOES NOT `assert()` that the return from
`Threads::owning_thread_from_monitor_owner()` is not NULL and in fact
has a comment that says:
```
    // owning_thread_from_monitor_owner() may also return NULL here
```

If memory serves, `ObjectSynchronizer::get_lock_owner()` can be called from
locations where we have non-NULL `owner`, but when we try to find the owning
thread, we can sometimes get a NULL back.